fml/raster_thread_merger_unittests.cc

```diff
@@ -616,11 +487,148 @@ TEST(RasterThreadMerger, SetMergeUnmergeCallback) {
+  ASSERT_FALSE(raster_thread_merger2_->IsMerged());
+}
+
+TEST(RasterThreadMerger, TheLastCallerOfMultipleMergersCanUnmergeNow) {
+  TaskQueueWrapper queue1;
+  TaskQueueWrapper queue2;
+  fml::TaskQueueId qid1 = queue1.GetTaskQueueId();
+  fml::TaskQueueId qid2 = queue2.GetTaskQueueId();
+  const auto raster_thread_merger1_ =
+      fml::RasterThreadMerger::CreateOrShareThreadMerger(nullptr, qid1, qid2);
+  const auto raster_thread_merger2_ =
+      fml::RasterThreadMerger::CreateOrShareThreadMerger(raster_thread_merger1_,
+                                                         qid1, qid2);
+  const int kNumFramesMerged = 5;
+  ASSERT_FALSE(raster_thread_merger1_->IsMerged());
+  ASSERT_FALSE(raster_thread_merger2_->IsMerged());
+
+  // Merge using the mergers
+  raster_thread_merger1_->MergeWithLease(kNumFramesMerged);
+  ASSERT_TRUE(raster_thread_merger1_->IsMerged());
+  ASSERT_TRUE(raster_thread_merger2_->IsMerged());
+  // Extend the second merger's lease
+  raster_thread_merger2_->ExtendLeaseTo(kNumFramesMerged);
+  ASSERT_TRUE(raster_thread_merger1_->IsMerged());
+  ASSERT_TRUE(raster_thread_merger2_->IsMerged());
+
+  // Two callers state becomes one caller left.
+  raster_thread_merger1_->UnMergeNowIfLastOne();
+  // Check if still merged
+  ASSERT_TRUE(raster_thread_merger1_->IsMerged());
+  ASSERT_TRUE(raster_thread_merger2_->IsMerged());
+
+  // One caller state becomes no callers left.
+  raster_thread_merger2_->UnMergeNowIfLastOne();
+  // Check if unmerged
+  ASSERT_FALSE(raster_thread_merger1_->IsMerged());
+  ASSERT_FALSE(raster_thread_merger2_->IsMerged());
+}
+
+TEST(RasterThreadMerger,
+     MultipleMergersCanIndependentlyMergeTwoOrMoreThreadsIntoOne) {
+  TaskQueueWrapper queue1;
+  TaskQueueWrapper queue2;
+  TaskQueueWrapper queue3;
+  fml::TaskQueueId qid1 = queue1.GetTaskQueueId();
+  fml::TaskQueueId qid2 = queue2.GetTaskQueueId();
+  fml::TaskQueueId qid3 = queue3.GetTaskQueueId();
+
+  const auto raster_thread_merger1_ =
+      fml::RasterThreadMerger::CreateOrShareThreadMerger(nullptr, qid1, qid2);
+  const auto raster_thread_merger2_ =
+      fml::RasterThreadMerger::CreateOrShareThreadMerger(raster_thread_merger1_,
+                                                         qid1, qid3);
+  const int kNumFramesMerged = 5;
+  ASSERT_FALSE(raster_thread_merger1_->IsMerged());
+  ASSERT_FALSE(raster_thread_merger2_->IsMerged());
+
+  // Merge thread2 into thread1
+  raster_thread_merger1_->MergeWithLease(kNumFramesMerged);
+  // Merge thread3 into thread1
+  raster_thread_merger2_->MergeWithLease(kNumFramesMerged);
+
+  ASSERT_TRUE(raster_thread_merger1_->IsMerged());
+  ASSERT_TRUE(raster_thread_merger2_->IsMerged());
+
+  for (int i = 0; i < kNumFramesMerged; i++) {
+    ASSERT_TRUE(raster_thread_merger1_->IsMerged());
+    raster_thread_merger1_->DecrementLease();
+  }
+
+  ASSERT_FALSE(raster_thread_merger1_->IsMerged());
+  ASSERT_TRUE(raster_thread_merger2_->IsMerged());
```

Comment:
What does this mean that one of them is merged but not the other?

Reply:
This case is different from the others cases above.  It tests the triple thread merging case: merger1 does the merging job from raster1 to platform1, and merger2 does the merging job from raster2 to platform1.
Maybe I can change a better test case name.